### ISSUE TITLE
Real OpenGL on Raspberry Pi

### ISSUE THREAD

**Issue by troutsneeze** [liballeg/allegro5#745]
Raspberry Pi 2/3 now support real OpenGL in a window etc. It's in development and slow but may be worth supporting at some point.

**Comment by dos1** [comment]
I've been already using Allegro on Raspberry Pi with real OpenGL by just building it on-device just like on a standard GNU/Linux distribution (as opposed to using Toolchain-raspberrypi.cmake file) and it seemed to work just fine. Is there anything else needed?

**Comment by SiegeLord** [comment]
The key aspect, I think, is the 'in a window' bit. Right now, you can only have fullscreen Allegro programs, while in principle there is no reason why we can't have it work like any other Linuxes.

**Comment by dos1** [comment]
Yes, I got it, and that's what I meant. When I build it without any Raspberry Pi-specific toolchain while having experimental OpenGL driver enabled, Allegro works just like on a regular distribution - in a standard X11 window.

**Comment by SiegeLord** [comment]
Oh I see, fascinating. I guess it needs more investigation to see if there's still performance benefits to our dedicated RaspberryPi code. Either way, we can mention this in the readme.

**Comment by dos1** [comment]
The catch is - dedicated code is necessary when using standard driver, and standard Linux code is necessary when using experimental driver; they aren't mutually compatible :)

**Comment by goobliata** [comment]
I thought it was using software rendering if you built the standard Linux way. @anholt recently made the experimental driver that allows hardware rendering on Pi 2 and up.

**Comment by goobliata** [comment]
By "slow" in the original post I didn't mean as slow as software, just slower than the way we currently do fullscreen.

**Comment by goobliata** [comment]
Oh I guess I see what you're saying @dos1. Yeah, what's needed is a way to switch between the two, maybe al_set_pi_display_driver(bool vc4) or something like that.

**Comment by goobliata** [comment]
Yeah, I added this ticket because the expermental driver is new.

**Comment by dos1** [comment]
You cannot have both drivers active on Raspberry Pi at the same time anyway, so all Allegro would need to have is just a single build that supports both :)

**Comment by goobliata** [comment]
That's true, it would just have to detect it somehow (reading the pi settings I guess.)

**Comment by dos1** [comment]
Just a note: I'll return to it later and that's not a scientific benchmark at all, but I've played a bit with Allegro on my Raspberry Pi and noticed that standard linuxy build (with GLES enabled though, see #826) on the new open driver was way faster than the Raspberry Pi-specific build on the old proprietary GLES drivers.

However, turns out that this new open driver is not supported on Raspberry Pi 0 and 1 (at least not officially).

**Comment by dos1** [comment]
This could be an issue though: https://jogamp.org/bugzilla/show_bug.cgi?id=1290

**Comment by dos1** [comment]
Proprietary GLES drivers have changed names in Raspbian Stretch: https://www.raspberrypi.org/forums/viewtopic.php?t=191857#p1203490

**Comment by troutsneeze** [comment]
I still haven't built Allegro this way, but I've built SDL this way. I don't even have any Pi anymore to benchmark the difference. I noticed as the standard OpenGL driver on Pi became mainstream, the old Allegro code became slower. Maybe it's up to speed now. But I'm not sure if Pi 0/1 are supported yet. One of my games which runs on those would never run under X11 though, I need to set the memory split to 128 MB and it needs every bit of both. But I'm not super concerned with the Pi anymore. But if the driver I wrote is still working and supports otherwise unsupported devices with Allegro's full feature set, then I see no point in removing it. But this issue can be closed I think.
